lib/stdlib/src/timer.erl

```diff
@@ -26,129 +26,192 @@
 	 cancel/1, sleep/1, tc/1, tc/2, tc/3, now_diff/2,
 	 seconds/1, minutes/1, hours/1, hms/3]).
 
--export([start_link/0, start/0, 
-	 handle_call/3,  handle_info/2,  
+-export([start_link/0, start/0,
+	 handle_call/3,  handle_info/2,
 	 init/1,
 	 code_change/3, handle_cast/2, terminate/2]).
 
-%% internal exports for test purposes only
--export([get_status/0]).
-
 %% types which can be used by other modules
 -export_type([tref/0]).
 
-%% Max
--define(MAX_TIMEOUT, 16#0800000).
--define(TIMER_TAB, timer_tab).
--define(INTERVAL_TAB, timer_interval_tab).
+%% Max value for a receive's after clause.
+-define(MAX_RECEIVE_AFTER, 16#ffffffff).
+
+%% validations
+-define(valid_time(T), is_integer(T), T >= 0).
+-define(valid_mfa(M, F, A), is_atom(M), is_atom(F), is_list(A)).
 
 %%
 %% Time is in milliseconds.
 %%
--opaque tref()    :: {integer(), reference()}.
+-opaque tref()    :: {type(), reference()}.
+-type type()      :: 'once' | 'interval'.
 -type time()      :: non_neg_integer().
 
 %%
 %% Interface functions
 %%
 -spec apply_after(Time, Module, Function, Arguments) ->
-                         {'ok', TRef} | {'error', Reason} when
-      Time :: time(),
-      Module :: module(),
-      Function :: atom(),
-      Arguments :: [term()],
-      TRef :: tref(),
-      Reason :: term().
-
-apply_after(Time, M, F, A) ->
-    req(apply_after, {Time, {M, F, A}}).
-
--spec send_after(Time, Pid, Message) -> {'ok', TRef} | {'error', Reason} when
-      Time :: time(),
-      Pid :: pid() | (RegName :: atom()),
-      Message :: term(),
-      TRef :: tref(),
-      Reason :: term().
-send_after(Time, Pid, Message) ->
-    req(apply_after, {Time, {?MODULE, send, [Pid, Message]}}).
-
--spec send_after(Time, Message) -> {'ok', TRef} | {'error', Reason} when
-      Time :: time(),
-      Message :: term(),
-      TRef :: tref(),
-      Reason :: term().
+	    {'ok', TRef} | {'error', Reason}
+  when Time :: time(),
+       Module :: module(),
+       Function :: atom(),
+       Arguments :: [term()],
+       TRef :: tref(),
+       Reason :: term().
+apply_after(0, M, F, A)
+  when ?valid_mfa(M, F, A) ->
+    do_apply({M, F, A});
```

Comment:
The return value will not be correct:

```
1> timer:apply_after(0, erlang, abs, [-42]).
<0.85.0>
2> timer:apply_after(0, erlang, exit, [[], whatever]).
{'EXIT',{badarg,[{erlang,exit,
                         [undefined,whatever],
                         [{error_info,#{module => erl_erts_errors}}]},
                 {timer,do_apply,1,[{file,"timer.erl"},{line,468}]},
                 {erl_eval,do_apply,6,[{file,"erl_eval.erl"},{line,689}]},
                 {shell,exprs,7,[{file,"shell.erl"},{line,686}]},
                 {shell,eval_exprs,7,[{file,"shell.erl"},{line,642}]},
                 {shell,eval_loop,3,[{file,"shell.erl"},{line,627}]}]}}
```

Please update the test suite to test the return values.

(The reason that Dialyzer did not complain is probably because `do_apply/1` uses an old-style `catch`. An old-style `catch` can hide type violations for Dialyzer.)

Reply:
Funny thing is, I believe we had this right in one of the earlier commits =^^=

Reply:
Fixed and added return values to tests in latest commit.